src/util/timer.rs

```diff
@@ -0,0 +1,216 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::mem::replace;
+use std::cmp::{Ord, Ordering};
+use std::time::Duration;
+use std::collections::BinaryHeap;
+
+use util::time::Instant;
+
+pub struct Timer<T> {
+    pending: BinaryHeap<TimeoutTask<T>>,
+}
+
+impl<T> Timer<T> {
+    pub fn new(capacity: usize) -> Self {
+        Timer {
+            pending: BinaryHeap::with_capacity(capacity),
+        }
+    }
+
+    /// Add a periodic task into the `Timer`.
+    pub fn add_task(&mut self, timeout: Duration, task: T) {
+        let task = TimeoutTask {
+            next_tick: Instant::now() + timeout,
+            task: task,
+        };
+        self.pending.push(task);
+    }
+
+    /// Remove a task from the timer. Returns the `TimeoutTask` if found.
+    ///
+    /// If there are many tasks according with the condition, only the first
+    /// task will be removed from pending queue.
+    pub fn remove_task<F>(&mut self, f: F) -> Option<T>
```

Comment:
From what I can see, this method won't be used. If it's needed in the future, it can be added back at that time.